compiler/noirc_frontend/src/ownership/mod.rs

```diff
@@ -0,0 +1,502 @@
+        } else {
+            Some(parameter)
+        };
+
+        recur_on_parameter(
+            parameter,
+            parameter_type,
+            name,
+            &mut seen_array_types,
+            &mut new_bindings,
+        );
+    }
+
+    new_bindings
+}
+
+/// Recur on a parameter's type, digging into any struct fields, looking for references to arrays.
+/// This will build up an Expression of the current parameter access we're doing, e.g. `*foo.bar`
+/// would correspond to a parameter `foo` with struct field `bar` that is a reference to an array.
+///
+/// This function inserts a .clone() expression to any parameter arrays behind references with
+/// repeated types since these may potentially be aliased by other parameters.
+///
+/// If `parameter` is `None` we'll still traverse the type to find any array types mentioned but we
+/// will not issue any clones. This is required e.g. on a mutable by-value parameter like `mut x: ...`
+/// since `x` may contain arrays internally that we'll need to remember in case there is another
+/// parameter which uses them. E.g. `mut x: [Field; 2], y: &mut [Field; 2]`.
+fn recur_on_parameter<'typ>(
+    parameter: Option<Expression>,
+    parameter_type: &'typ Type,
+    parameter_name: &str,
+    seen_array_types: &mut HashSet<&'typ Type>,
+    new_bindings: &mut Vec<(String, Type, Expression)>,
+) {
+    match parameter_type {
+        // These types never contain arrays
+        Type::Field | Type::Integer(..) | Type::Bool | Type::Unit | Type::Function(..) => (),
+
+        Type::Array(_, element_type) | Type::Slice(element_type) => {
+            seen_array_types.insert(parameter_type);
+            // Don't clone inside arrays, but still look for more array types
+            recur_on_parameter(None, element_type, parameter_name, seen_array_types, new_bindings);
+        }
+        Type::String(_) | Type::FmtString(..) => {
+            seen_array_types.insert(parameter_type);
+        }
+
+        Type::Tuple(fields) => {
+            for (i, field) in fields.iter().enumerate() {
+                let expr = parameter.clone().map(|p| Expression::ExtractTupleField(Box::new(p), i));
+                recur_on_parameter(expr, field, parameter_name, seen_array_types, new_bindings);
+            }
+        }
+
+        Type::Reference(element_type) => {
+            if let Some(array) = inner_array_or_slice_type(element_type) {
+                if let Some(parameter) = parameter {
+                    // Check if the parameter type has been seen before
+                    if !seen_array_types.insert(array) {
+                        let typ = element_type.as_ref().clone();
+                        let expr = Expression::Unary(Unary {
+                            operator: UnaryOp::Dereference { implicitly_added: true },
+                            rhs: Box::new(parameter),
+                            result_type: typ.clone(),
+                            location: Location::dummy(), // TODO
```

Comment:
I don't know if this TODO can be fixed, maybe it can be set to the parameter's location if we have it?